### PR TITLE
[#92205218] lockmode

### PR DESCRIPTION
**Pull request by ramen74**
Initial crack at Read-Only access for pull team members.
Comment and Lock Functionality. 
Some other cleanup.
LOCK UI is a bit clunky now.

### CODE REVIEW

File: templates/index.html [hunk 1]
```diff
@@ -56,6 +60,11 @@
 
   <script type="text/javascript">
 
+  // initialize bootstrap tooltip 
+  $(function () {
+    $('[data-toggle="tooltip"]').tooltip()
```
Comment (kelvintaywl) [inline templates/index.html]: I think we can place the `$('[data-toggle="tooltip"]').tooltip();` line in the `$(function () {` function / [IIFE](http://en.wikipedia.org/wiki/Immediately-invoked_function_expression)  below. They are / can be executed altogether :+1:

File: templates/base.html [hunk 2]
```diff
@@ -5,8 +5,9 @@
   <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
   <link href="//netdna.bootstrapcdn.com/bootstrap/3.0.0/css/bootstrap.min.css" rel="stylesheet">
   <link href='//fonts.googleapis.com/css?family=Roboto+Condensed' rel='stylesheet' type='text/css'>
-  <link rel="shortcut icon" href="/static/favicon.ico">
-  <script type="text/javascript" src="//ajax.googleapis.com/ajax/libs/jquery/1.10.2/jquery.min.js"></script>
+  <link href='/static/css?family=Roboto+Condensed' rel='stylesheet' type='text/css'>
```
Comment (kelvintaywl) [inline templates/base.html]: this line seems unnecessary. perhaps a typo?

File: goship.go [hunk 3]
```diff
@@ -613,6 +628,55 @@ func DeployHandler(w http.ResponseWriter, r *http.Request) {
...
+	if err != nil {
+		log.Println("ERROR: ", err)
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	http.Redirect(w, r, "/", http.StatusMovedPermanently)
+}
+
+// UnLockHandler allows you to unlock an environment
+// http://127.0.0.1:8000/lock?environment=staging&project=admin
```
Comment (kelvintaywl) [inline goship.go]: `lock` -> `unlock` perhaps?

```
// http://127.0.0.1:8000/unlock?environment=staging&project=admin
```

File: goship.go [hunk 4]
```diff
@@ -166,7 +168,12 @@ func retrieveCommits(project goship.Project, deployUser string) goship.Project {
 			project.Environments[i].Hosts[j] = host
 		}
 	}
-	return project
+	u, err := getUser(r)
+	if err != nil {
+		log.Printf("Failed to get user %s", err)
+	}
+	proj := filterProject(project, r, u)
```
Comment (yugui) [inline goship.go]: Do you need this "proj" variable?

File: goship.go [hunk 5]
```diff
@@ -332,13 +339,14 @@ func DeployOutputHandler(w http.ResponseWriter, r *http.Request, env string, for
 	w.Write(b)
 }
 
-func DeployLogHandler(w http.ResponseWriter, r *http.Request, full_env string, environment goship.Environment) {
+// DeployLogHandler shows data abut the environment inclusing the deploy log.
```
Comment (yugui) [inline goship.go]: typo. "s/abut/about"

File: goship.go [hunk 6]
```diff
@@ -166,7 +168,12 @@ func retrieveCommits(project goship.Project, deployUser string) goship.Project {
 			project.Environments[i].Hosts[j] = host
 		}
 	}
-	return project
+	u, err := getUser(r)
+	if err != nil {
+		log.Printf("Failed to get user %s", err)
```
Comment (yugui) [inline goship.go]: Return err to the caller

File: goship.go [hunk 7]
```diff
@@ -613,6 +628,55 @@ func DeployHandler(w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+	http.Redirect(w, r, "/", http.StatusFound)
+
+}
+
+// LockHandler allows you to lock an environment
+// http://127.0.0.1:8000/lock?environment=staging&project=admin
+func LockHandler(w http.ResponseWriter, r *http.Request) {
+
```
Comment (yugui) [inline goship.go]: Remove this blank line

File: goship.go [hunk 8]
```diff
@@ -613,6 +628,55 @@ func DeployHandler(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+// CommentHandler allows you to update a comment on an environment
+// i.e. http://127.0.0.1:8000/comment?environment=staging&project=admin&comment=DONOTDEPLOYPLEASE!
+func CommentHandler(w http.ResponseWriter, r *http.Request) {
+
```
Comment (yugui) [inline goship.go]: Remove this blank line

File: goship.go [hunk 9]
```diff
@@ -613,6 +628,55 @@ func DeployHandler(w http.ResponseWriter, r *http.Request) {
...
+	env := r.FormValue("environment")
+	comment := r.FormValue("comment")
+	err := goship.SetComment(c, p, env, comment)
+	if err != nil {
+		log.Println("ERROR: ", err)
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	http.Redirect(w, r, "/", http.StatusFound)
+
```
Comment (yugui) [inline goship.go]: Remove blank line

**Comment by hayate** [comment]
:+1:

File: goship.go [hunk 10]
```diff
@@ -613,6 +628,55 @@ func DeployHandler(w http.ResponseWriter, r *http.Request) {
...
+	p := r.FormValue("project")
+	env := r.FormValue("environment")
+	comment := r.FormValue("comment")
+	err := goship.SetComment(c, p, env, comment)
+	if err != nil {
+		log.Println("ERROR: ", err)
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	http.Redirect(w, r, "/", http.StatusFound)
```
Comment (yugui) [inline goship.go]: StatusSeeOther looks to be more appropriate

File: goship.go [hunk 11]
```diff
@@ -613,6 +628,55 @@ func DeployHandler(w http.ResponseWriter, r *http.Request) {
...
+	c := etcd.NewClient([]string{*ETCDServer})
+	p := r.FormValue("project")
+	env := r.FormValue("environment")
+	err := goship.LockEnvironment(c, p, env, "false")
+	if err != nil {
+		log.Println("ERROR: ", err)
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	http.Redirect(w, r, "/", http.StatusMovedPermanently)
```
Comment (yugui) [inline goship.go]: StatusSeeOther looks to be more appropriate

File: goship.go [hunk 12]
```diff
@@ -613,6 +628,55 @@ func DeployHandler(w http.ResponseWriter, r *http.Request) {
...
+	c := etcd.NewClient([]string{*ETCDServer})
+	p := r.FormValue("project")
+	env := r.FormValue("environment")
+	err := goship.LockEnvironment(c, p, env, "true")
+	if err != nil {
+		log.Println("ERROR: ", err)
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	http.Redirect(w, r, "/", http.StatusMovedPermanently)
```
Comment (yugui) [inline goship.go]: StatusSeeOther looks to be more appropriate

File: goship.go [hunk 13]
```diff
@@ -977,22 +1070,80 @@ func getAuth() auth {
...
-	m, _, err := c.Repositories.IsCollaborator(owner, repo, user)
+	return githubClientProd{
+		org:  c.Organizations,
+		repo: c.Repositories,
+	}
+}
+
+// Will return true if the user has a team permission non read only
+func userHasDeployPermission(g githubClient, owner, repo, user string) (pull bool, err error) {
+
```
Comment (yugui) [inline goship.go]: remove blank line.

File: goship.go [hunk 14]
```diff
@@ -836,20 +904,52 @@ func checkAuth(fn http.HandlerFunc, a auth) http.HandlerFunc {
...
+			}
+			p.Environments[i].IsLocked = !lock
+			// Add a line break if there is already a comment
+			if p.Environments[i].Comment != "" {
+				p.Environments[i].Comment = p.Environments[i].Comment + " | "
+			}
+			if !lock {
+				p.Environments[i].Comment = p.Environments[i].Comment + "you do not have permission to deploy "
+			}
+		} else {
```
Comment (yugui) [inline goship.go]: golang prefers less indent for normal flows.

probably the "else" clause is the more common flow in this function?
Then continue at the end of the "if" clause and dedent the "else" clause.

File: goship.go (hunk 5 above)
Comment (kelvintaywl) [inline goship.go]: ah, there's another typo:  `inclusing` -> `including` i guess

**Pull request merged**